### PR TITLE
Mimecast

### PR DESCRIPTION
**Pull request by eyalpalo**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/projects/CIAC/issues/CIAC-1389?filter=myopenissues)

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 1]
```diff
@@ -358,7 +819,7 @@ def get_policy():
     title = 'Mimecast list blocked sender policies: \n These are the existing Blocked Sender Policies:'
     policy_id = demisto.args().get('policyID')
     if policy_id:
-        policy_id = policy_id.encode('utf-8')
+        policy_id = policy_id
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:822]:
```suggestion
```

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 2]
```diff
@@ -556,7 +1017,7 @@ def set_empty_value_args_policy_update(policy_obj, option, policy_id):
      """
     empty_args_list = []
     # Add the empty arguments to empty args list
-    for arg, value in policy_obj.items():
+    for arg, value in list(policy_obj.items()):
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:1020]:
```suggestion
    for arg, value in policy_obj.items():
```
I guess 2to3 added this, I think it's redundant

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 3]
```diff
@@ -772,7 +1233,7 @@ def list_managed_url():
     full_url_response = ''
     url = demisto.args().get('url')
     if url:
-        url = url.encode('utf-8')
+        url = url
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:1236]: remove this if

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 4]
```diff
@@ -823,25 +1284,25 @@ def list_managed_url_request():
...
+    url = demisto.args().get('url')
+    action = demisto.args().get('action')
+    match_type = demisto.args().get('matchType')
+    disable_rewrite = demisto.args().get('disableRewrite')
+    disable_user_awareness = demisto.args().get('disableUserAwareness')
+    disable_log_click = demisto.args().get('disableLogClick')
     comment = demisto.args().get('comment')
     if comment:
-        comment = comment.encode('utf-8')
+        comment = comment
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:1305]: same

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 5]
```diff
@@ -950,44 +1418,39 @@ def list_messages():
...
 
     if from_date:
         search_params['from'] = from_date
     if to_date:
         search_params['to'] = to_date
     if scan_result:
         search_params['scanResult'] = scan_result
-
-    url_logs = get_url_logs_request(search_params, limit=limit)
+    demisto.results('here')
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:1447]:
```suggestion
```

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 6]
```diff
@@ -2308,14 +2705,257 @@ def search_file_hash_api_response_to_context(api_response):
...
+    response = http_request('POST', api_endpoint='/api/gateway/get-hold-summary-list', payload={'data': []})
+    if response.get('fail'):
+        raise Exception(json.dumps(response.get('fail')[0].get('errors')))
+
+    summery_list = response.get('data')
+
+    headers = {'policyInfo': 'Held Reason',
+               'numberOfItems': 'Number Of Items'
+               }
+    readable_output = tableToMarkdown('Message Summery', t=summery_list,
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:2764]:
```suggestion
    readable_output = tableToMarkdown('Message Summary', t=summery_list,
```

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 7]
```diff
@@ -2308,14 +2705,257 @@ def search_file_hash_api_response_to_context(api_response):
...
+
+    """
+    response = reject_hold_message_request(args)
+    readable_output = ''
+
+    if response.get('fail'):
+        raise Exception(json.dumps(response.get('fail')[0].get('errors')))
+    for message in response.get('data', []):
+        if not message.get('reject', False):
+            readable_output += f'Hold message with id {message.get("id")} rejection failed.\n'
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:2846]: why not fail the command in this case? when will it be false?

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 8]
```diff
@@ -2308,14 +2705,257 @@ def search_file_hash_api_response_to_context(api_response):
...
+            args: input arguments for the command.
+
+        """
+    id = args.get('id')
+    response = release_hold_message_request(id)
+
+    if response.get('fail'):
+        raise Exception(json.dumps(response.get('fail')[0].get('errors')))
+    if not response.get('data', [])[0].get('release', False):
+        readable_output = 'Message release has failed.'
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:2870]: same here

**Review by dantavori — CHANGES_REQUESTED** [review]
lmk when to review the yml.
I'll complete the review (mainly for the tests) after the demo.

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py [hunk 9]
```diff
@@ -1345,6 +1760,14 @@ def impersonation_to_incident(impersonation_log):
     return incident
 
 
+def held_to_incident(held_message):
+    incident = {}
+    incident['name'] = f'Mimecast held message: {held_message.get("subject")}'
+    incident['occurred'] = held_message.get('dateReceived').replace('+0000', 'Z')
+    incident['rawJSON'] = json.dumps(held_message)
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.py:1767]: let's add dbotMirrorId as described here https://xsoar.pan.dev/docs/integrations/fetching-incidents#fetch-history

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 10]
```diff
@@ -1631,15 +1748,802 @@ script:
...
+      type: Number
+    - contextPath: Mimecast.HeldMessageSummary.policyInfo
+      description: The name of the policy or definition that held a message.
+      type: String
+  - name: mimecast-reject-held-message
+    description: Reject a currently held message.
+    deprecated: false
+    arguments:
+    - name: ids
+      description: "An array of Mimecast secure ids: Ids are extracted from command : mimecast-list-held-message"
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:2332]:
```suggestion
      description: "A comma-separated list of Mimecast secure ids: Ids are extracted from command : mimecast-list-held-message"
```

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 11]
```diff
@@ -1631,15 +1748,802 @@ script:
...
+      - ""
+    - name: sender_ip
+      description: "Part of advancedTrackAndTraceOptions object : The source IP address of messages to track."
+      required: false
+      isArray: false
+      defaultValue: ""
+      predefined:
+      - ""
+    - name: route
+      description: An array of routes to filter by. Possible values are internal, outbound and inbound.
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:1812]:
```suggestion
      description: A comma-separated list of routes to filter by. Possible values are internal, outbound and inbound.
```
let's check if `Possible values are internal, outbound and inbound.` is not auto genetated in the UI

File: Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml [hunk 12]
```diff
@@ -1631,15 +1748,802 @@ script:
...
+      type: Number
+    - contextPath: Mimecast.SearchMessage.detectionLevel
+      description: Detection level of the email.
+      type: String
+  - name: mimecast-get-message-info
+    description: Retrieves detailed information about a specific message.
+    deprecated: false
+    arguments:
+    - name: ids
+      description: The Mimecast ID of the message to load. This is returned by the /api/message-finder/search endpoint. (mimecast-search-message command)
```
Comment (dantavori) [inline Packs/Mimecast/Integrations/MimecastV2/MimecastV2.yml:1872]:
```suggestion
      description: A comma-separated list of the Mimecast IDs of the messages to load. This is returned by the /api/message-finder/search endpoint. (mimecast-search-message command)
```

**Review by dantavori — APPROVED** [review]
Looks great, see last comments from my side.
please review the README: parameters and mimecast-reject-held-message sections are broken.
Please assign docs team to go over the yml file and then regenerate the readme.

**Comment by eyalpalo** [comment]
@ShirleyDenkberg Can you have a look at the yaml file for doc review?

**Comment by ShirleyDenkberg** [comment]
@eyalpalo I reviewed the yml file and have the following questions:
Line 136, is there a spelling mistake?  What is fomr?  What is this supposed to say?
 description: Search in specific dates only (default is all mails **fomr**)
Lines 846 and 847 - is default **all** or **malicious**? The description and defaultValue do not match.
      defaultValue: **malicious**
      description: Filters logs by scan result, default is **all**
Lines 2280 - 2284  Is the display name **from envelope** or **from header**?  The context path says from header.
   - contextPath: Mimecast.HeldMessage.**fromHeader**.displayableName
      description: The display name of the sender (**From envelope**).
      type: String
    - contextPath: Mimecast.HeldMessage.**fromHeader**.emailAddress
      description: The email address of the sender (**From envelope**).
Line 2427 Don't understand this description.  Do you need to enter the attachment or are you supposed to indicate (true/false) if there is an attachment? (In the latter case, the description should be **Whether there is an attachment in the message.**)
name: attachments
      description: The presence of an attachment in the message.

When these issues are resolved (updated) the readme needs to be regenerated.

**Comment by ShirleyDenkberg** [comment]
@dantavori Doc review completed.

**Pull request merged**